### DISCUSSION TITLE
Battery pre-condition

### DISCUSSION THREAD

**Discussion by deepanshu-shukla** [BottlecapDave/HomeAssistant-OctopusEnergy#1053] [Ideas]
It may sound weird but GivEnergy batteries have a throttle that reduces the charging speed to 2.7KwH instead of full 3.5KwH when the battery cell temp falls below 20° C. 

This means that if we have selected 3Hr to charge the battery, it wont be sufficient anymore. 

A workaround as suggested by a forum member was to discharge or charge for 5-10 mins prior to main slot charging so that battery temp will be raised and full capacity charging can take place. 

One solution of implementing this would be to have a "Battery Precondition " sensor and a "Battery Precondition Time" that can be configured by the user in minutes, say "X". The sensor would would come on at T-X and off at T-1 where T is the slot time.

**Comment by BottlecapDave** [comment]
Hello. Target rate sensors already support [offsets](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy/setup/target_rate/#offset) which will adjust when the target rate sensor comes on relative to the start time of the selected rate period.

**Reply by deepanshu-shukla** [reply]
yes, but that's not helpful in this scenario. When you want to "precondition" the battery then it should be warm enough before the target slot so that it charges at peak value of 3.5KwH else if it starts when temp is low, it will continue charging at 2.7Kwh. 

If we have an additional offset sensor that would come on and off before the actual slot, then we can use that to charge or discharge so that cell temp is raised.

**Reply by BottlecapDave** [reply]
The offset turns the binary sensor on before/after the discovered target slots. So if you have an offset set to -00:30:00 and a target time that starts at 10:00:00, then the sensor will turn on at 09:30:00. You can then precondition your battery when the sensor turns on. 

Alternatively, you can setup a template sensor which looks at the first discovered target time. The following _should_ give you a binary sensor that turns on 30 minutes before the first target time

```
{% set times = state_attr('binary_sensor.octopus_energy_target_battery', 'target_times') %}
{% if times != null and times != none and times | length > 0 %}
{{ now() >= (times[0]['start'] - timedelta(minutes=30)) and now() <= (times[0]['end'] - timedelta(minutes=30)) }}
{% else %}
false
{%endif%}
```
